staging/src/k8s.io/apiserver/pkg/endpoints/request/requestinfo.go

```diff
@@ -216,6 +219,22 @@ func (r *RequestInfoFactory) NewRequestInfo(req *http.Request) (*RequestInfo, er
 		} else {
 			requestInfo.Verb = "list"
 		}
+
+		// For list/watch requests, if there is a fieldSelector set and enforcing
```

Comment:
is there a reason not to decode the query the same way as the list handler? that would unify what we're doing for `watch` and `fieldSelector`, and would make me more confident all the edge cases around multiple query parameters or decoding oddities would be handled identically.

```
if len(requestInfo.Name) == 0 && requestInfo.Verb == "get" {
  opts := metainternalversion.ListOptions{}
  if err := metainternalversion.ParameterCodec.DecodeParameters(req.URL.Query(), scope.MetaGroupVersion, &opts); err != nil {
    return err
  }

  if opts.Watch {
    requestInfo.Verb = "watch"
  } else {
    requestInfo.Verb = "list"
  }

  if opts.FieldSelector != nil {
    if name, ok := opts.FieldSelector.RequiresExactMatch("metadata.name"); ok {
      requestInfo.Name = name
    }
  }
}
```

Reply:
It was a bit more complex to handle tests correctly, but done now.